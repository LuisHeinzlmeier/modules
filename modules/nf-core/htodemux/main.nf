process HTODEMUX {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/r-seurat_r-seuratobject:4c5a804804327d29':
        'community.wave.seqera.io/library/r-seurat_r-seuratobject:b11306d1bdc82827' }"

    input:
    tuple val(meta), path(seurat_object), val(assay)

    output:
    tuple val(meta), path("*_params_htodemux.csv")        , emit: params
    tuple val(meta), path("*_assignment_htodemux.csv")    , emit: assignment
    tuple val(meta), path("*_classification_htodemux.csv"), emit: classification
    tuple val(meta), path("*_htodemux.rds")               , emit: rds
    path "versions.yml"                                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    quantile = task.ext.quantile ?: 0.99         // The quantile of inferred 'negative' distribution for each hashtag, over which the cell is considered positive
    init     = task.ext.init     ?: null         // Initial number of clusters for hashtags
    nstarts  = task.ext.nstarts  ?: 100          // nstarts value for k-means clustering
    kfunc    = task.ext.kfunc    ?: "clara"      // Clustering function for initial hashtag grouping. Default is clara for fast k-medoids clustering on large applications, also support kmeans for kmeans clustering
    nsamples = task.ext.nsamples ?: 100          // Number of samples to be drawn from the dataset used for clustering, for kfunc = clara
    seed     = task.ext.seed     ?: 42           // Sets the random seed
    verbose  = task.ext.verbose  ?: true         // Verbose output
    prefix   = task.ext.prefix   ?: "${meta.id}" // Prefix name for the output files

    template 'HTODemux.R'

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_params_htodemux.csv
    touch ${prefix}_assignment_htodemux.csv
    touch ${prefix}_classification_htodemux.csv
    touch ${prefix}_htodemux.rds

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(Rscript -e "cat(strsplit(R.version[['version.string']], ' ')[[1]][3])")
        r-seurat: \$(Rscript -e "library(Seurat); cat(as.character(packageVersion('Seurat')))")
    END_VERSIONS
    """
}
