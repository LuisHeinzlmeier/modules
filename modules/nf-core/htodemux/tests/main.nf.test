nextflow_process {

    name "Test Process HTODEMUX"
    script "../main.nf"
    process "HTODEMUX"

    tag "modules"
    tag "modules_nfcore"
    tag "htodemux"

    test("seuratObject - rds") {

        when {
            process {
                """
                // TODO nf-core: define inputs of the process here. Example:

                input[0] = [
                    [ id:'test'], // meta map
                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/htodemux.rds', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("seuratObject - rds - stub") {

        options "-stub"

        when {
            process {
                """
                // TODO nf-core: define inputs of the process here. Example:

                input[0] = [
                    [ id:'test'], // meta map
                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/htodemux.rds', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
}
