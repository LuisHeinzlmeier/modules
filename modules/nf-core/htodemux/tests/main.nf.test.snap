{
    "sarscov2 - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_assignment_htodemux.csv:md5,33d72b27c3d0b0f7394af18a18490a5c",
                            "test_classification_htodemux.csv:md5,886f595f148d2ccdd5a9134b8eb815cb",
                            "test_params_htodemux.csv:md5,61065a5ddedda3bcf5a51e9eb5a72eae"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "testhtodemux.rds:md5,ec43d2d0217c3a87529f1eb4d0e05ffb"
                    ]
                ],
                "2": [
                    "versions.yml:md5,faa097177f5c4efd711eb8dd5701b6d2"
                ],
                "csv": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_assignment_htodemux.csv:md5,33d72b27c3d0b0f7394af18a18490a5c",
                            "test_classification_htodemux.csv:md5,886f595f148d2ccdd5a9134b8eb815cb",
                            "test_params_htodemux.csv:md5,61065a5ddedda3bcf5a51e9eb5a72eae"
                        ]
                    ]
                ],
                "rds": [
                    [
                        {
                            "id": "test"
                        },
                        "testhtodemux.rds:md5,ec43d2d0217c3a87529f1eb4d0e05ffb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,faa097177f5c4efd711eb8dd5701b6d2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-14T15:21:00.420534"
    }
}